### ISSUE TITLE
Scrape English versions of ITU OB issues as .docx files

### ISSUE THREAD

**Issue by strogonoff** [ituob/itu-ob-data#16]
The goal is to write an utility that scrapes all ITU OB issues in .docx format (just English versions for now, and no annexes).

The issues are here: https://www.itu.int/en/publications/ITU-T/pages/publications.aspx?parent=T-SP-OB&version_date=2019

The site was recently down so we should probably throttle requests & be gentle on it.

The utility should store .docx files as `<issue ID>/<issue ID>-E.docx`, where issue ID is a simple integer like 1023.

Utility output should not be versioned. If we need to share the downloaded .docx file archive, we can upload it somewhere else.

**Comment by ronaldtse** [comment]
@strogonoff we should probably scrape all the languages available...?

**Comment by ronaldtse** [comment]
@andrew2net do you have time for this?

**Comment by strogonoff** [comment]
@ronaldtse I’d rather have English versions sooner than all languages later (and I don’t want us to accidentally DoS ITU’s site) so I vote to handle it incrementally

**Comment by strogonoff** [comment]
it’s obvious that we’ll need other languages eventually, but merging translations will be a challenge for later…

**Comment by ronaldtse** [comment]
Maybe Relaton-ITU should also provide the links for Word/PDF docs in English/other languages? Thoughts @andrew2net ?

Then this can be a simple wrapper script.

**Comment by strogonoff** [comment]
Updated issue description to add a note about annex handling, and updated the path structure for downloaded contents.

**Comment by strogonoff** [comment]
Note that OB IDs are sequential integers, but unfortunately each issue’s URL contains the year of that issue’s publication date (for example, https://www.itu.int/en/publications/ITU-T/pages/publications.aspx?parent=T-SP-OB.1165-2019). I don’t think we can reliably infer the year without actually going through the OB archive index year by year.

From what I saw in Relaton-ITU, it’s outside of what it’s intended to do (naturally), but perhaps there is a use for it somewhere…

**Comment by ronaldtse** [comment]
> Note that OB IDs are sequential integers, but unfortunately each issue’s URL also contains the year of that issue’s publication date (for example, https://www.itu.int/en/publications/ITU-T/pages/publications.aspx?parent=T-SP-OB.1165-2019). I don’t think we can reliably infer the year without actually going through OB archive pages year by year.

That's fine. Moreover, some OB issues have translations, and some not. So it is necessary to visit every page anyway.

> From what I saw in Relaton-ITU implementation, it’s outside of what it was intended to do (naturally), but perhaps there is a use for it somewhere…

It isn't -- Relaton-ITU is supposed to also provide accessible links for the document. DOC/PDF links are all in scope.

**Comment by andrew2net** [comment]
> @andrew2net do you have time for this?

@ronaldtse I have a bunch of uncompleted tasks. But of course, you can change prioritizing.

**Comment by strogonoff** [comment]
@ronaldtse

> some OB issues have translations, and some not

Didn’t know, that’s unfortunate.

> It isn't -- Relaton-ITU is supposed to also provide accessible links for the document. DOC/PDF links are all in scope.

If your point is that Relaton-ITU users may want to reference OB issues, then I can see how this may be in scope.

If we could give Relaton-ITU integer ID of OB issue and get document links in return, this would be easy. We can simply iterate over integers from 1 and gather links until it returns an error. Even if Relaton requires a year in addition to issue ID (OB issue URLs contain the year), that can be worked around.

That said, if Relaton-ITU doesn’t have ITU OB support yet, I believe it may be much faster to write a quick bespoke scraper just for this purpose.

**Comment by strogonoff** [comment]
This should be on hold until we sort out #20.

**Comment by strogonoff** [comment]
## Rough logic for scraping OB issues from itu.int

This does not handle issues older than 567, since they are not available through the same archive index on itu.int.

### Assumptions

* All .docx download URLs follow the following format:

    ```
    https://www.itu.int/dms_pub/itu-t/opb/sp/T-SP-OB.<integer OB ID>-<year>-[OAS-]<file format>-<one-letter language code>.<file extension>
    ```

    * The `OAS-` substring is present for older issues and not newer issues.
    * File formats we are interested in are `MSW` (stands for Microsoft Word) and `ZIP`.
    * File extension can be .doc, .docx (both with MSW) or .zip (with ZIP).

### Preliminary tasks

* Build a map of languages & one-letter codes. E.g., English is `E`.
* Determine the starting ITU OB & year (depends on #20—whether we have to process older issues in .doc or not)

### Logic

* Start with given OB ID and year (567 and 1994 if we choose to auto-process all old issues available on itu.int).
* Substitute variables in the URL format string & try downloading a docx for each language.
* If URL reports not found, try changing format from docx to zip; then from zip to doc. Try each with a combination of `OAS-` substring in the URL.
* If still not found, increment year by one & retry the probing again.
* If downloaded file is a .zip, extract it and determine which of the files is the issue itself and which are the annexes based on filename pattern (annexes usually have the word “annex” in file name, case-insensitive).

**Comment by strogonoff** [comment]
## Approach #2 

* Parse ITU OB archive webpages on ITU.int year by year and issue by issue.
* Build a data structure of download URLs. For each issue ID & language combination there will be one download URL (some languages may be missing for some issues).
* With that data structure as input, download all issues, expanding ZIP archives and placing/renaming files as needed.
